### PR TITLE
Allow Enumerable#pluck to take a splat.

### PR DESCRIPTION
**Pull request by kddnewton**
This allows easier integration with ActiveRecord, such that
AR#pluck will now use Enumerable#pluck if the relation is loaded,
without needing to hit the database.

### CODE REVIEW

**Comment by dhh** [comment]
Looking good to me!

**Comment by kddeisz** [comment]
Added then squashed one more commit because I didn't remember that there was a assert_no_difference

File: activerecord/test/cases/calculations_test.rb [hunk 1]
```diff
@@ -650,4 +664,10 @@ def test_should_reference_correct_aliases_while_joining_tables_of_has_many_throu
       developer.ratings.includes(comment: :post).where(posts: { id: 1 }).count
     end
   end
+  
+  private
+
+    def assert_no_relations_created(&block)
```
Comment (matthewd) [inline activerecord/test/cases/calculations_test.rb]: I think we should probably be asserting that no queries are run, instead

Comment (kddeisz) [inline activerecord/test/cases/calculations_test.rb]: Yeah didn't know how to do that. What would that look like?

**Comment by matthewd** [comment]
This breaks `Relation#pluck('arbitrary || sql || expression')`, where the relation is loaded.

In that situation, pretty much anything we could do is going to be less-than-ideal... but personally, I lean towards us running the query. (That is, I believe we should only apply the "just use the already-loaded data" behaviour if all the plucked items are simple column names.)

File: activerecord/lib/active_record/relation/calculations.rb [hunk 2]
```diff
@@ -153,23 +153,27 @@ def calculate(operation, column_name)
     # See also +ids+.
     #
     def pluck(*column_names)
-      column_names.map! do |column_name|
-        if column_name.is_a?(Symbol) && attribute_alias?(column_name)
-          attribute_alias(column_name)
-        else
-          column_name.to_s
+      if loaded?
+        @records.pluck(*column_names)
```
Comment (bogdan) [inline activerecord/lib/active_record/relation/calculations.rb]: Indent here flushes history and creates too much nesting when a lot of edge cases of a method handled at the beginning. Can we use return instead and leave indent as is?

File: activerecord/test/cases/calculations_test.rb (hunk 1 above)
Comment (kddeisz) [inline activerecord/test/cases/calculations_test.rb]: @matthewd Just fixed that. Thanks - didn't know about that assertion.

**Comment by dhh** [comment]
Ah, yes. Need to run the query in that case.

**Comment by kddeisz** [comment]
@matthewd - Sorry I'm going to need help on this then. How would you go about checking?

**Comment by matthewd** [comment]
It sounds like we can probably lean on [the existing conditional](https://github.com/rails/rails/blob/343dad9617e7448bf8effc7c70ae13bede38fd32/activerecord/lib/active_record/relation/calculations.rb#L157).

I'm not certain, but the `has_include?` bit sounds like something we might need to punt on too.

So, I'm thinking (something vaguely along the lines of) inserting a `too_hard = true` [here](https://github.com/rails/rails/blob/343dad9617e7448bf8effc7c70ae13bede38fd32/activerecord/lib/active_record/relation/calculations.rb#L160), and then `return @records.pluck(*column_names) if loaded? && !too_hard` [here](https://github.com/rails/rails/blob/343dad9617e7448bf8effc7c70ae13bede38fd32/activerecord/lib/active_record/relation/calculations.rb#L167).

Non-crappy variable naming left as an exercise for the reader :wink:

File: activerecord/lib/active_record/relation/calculations.rb (hunk 2 above)
Comment (kddeisz) [inline activerecord/lib/active_record/relation/calculations.rb]: Fixed.

**Comment by matthewd** [comment]
Actually.. that conditional isn't doing what I thought it was. So yeah, what you just did looks good. :+1:

**Comment by kddeisz** [comment]
Thanks!

File: activerecord/lib/active_record/relation/calculations.rb [hunk 3]
```diff
@@ -161,6 +161,10 @@ def pluck(*column_names)
         end
       end
 
+      if loaded? and (column_names - @klass.column_names).empty?
```
Comment (kaspth) [inline activerecord/lib/active_record/relation/calculations.rb]: Prefer `&&` here.

Comment (kddeisz) [inline activerecord/lib/active_record/relation/calculations.rb]: Fixed.

**Comment by kddeisz** [comment]
The only weirdness I still see is `klass.pluck(:id, :id)`, which will return duples, where AR will return just the single array. Should the enumerable be in charge of uniq, should AR, or should we ignore?

**Comment by dhh** [comment]
I don't think that case matters. I'd just ignore it.

**Comment by kddeisz** [comment]
Nice - then we're probably good to go unless someone else thinks of something.

**Comment by dhh** [comment]
Great work and feedback, everyone! Love how we flushed out all the edge and corner cases and got the work done so quickly. Viva la open source!
